File: nestMicro/clinical-micro/src/modules/clinicalRecord/controllers/clinicalRecord.controller.ts
import { Controller, Get, Post, Body, Param, Delete, ParseIntPipe, HttpCode, HttpStatus} from '@nestjs/common';
import { ApiTags, ApiOperation, ApiResponse, ApiParam } from '@nestjs/swagger';
import { ClinicalRecordService } from '../services/clinicalRecord.service';
import { CreateClinicalRecordDto } from '../dto/create-clinicalRecord';
import { ClinicalRecord } from '../entities/ClinicalRecord.entity';

@ApiTags('Tipos de Tumor')
@Controller('clinical-records')
export class ClinicalRecordController{
    constructor(private readonly clinicalRecordService: ClinicalRecordService){}


    @Post()
    @HttpCode(HttpStatus.CREATED)
    @ApiOperation({
        summary: 'Crea un nuevo registro clinico',
        description: 'Registra un nuevo diagnostico de tumor, etapa y protocolo de trtamiento para el paciente'

    })

    @ApiResponse({
        status: 201,
        description: 'Registro clinico creado exitosamente',
        type: ClinicalRecord
    })

    @ApiResponse({
        status:404,
        description:'Datos de entrada invalidos (IDs o formato de fecha incorrectos)'
    })
    
    
    create(@Body() createClinicalRecordDto: CreateClinicalRecordDto): Promise<ClinicalRecord>{
        return this.clinicalRecordService.create(createClinicalRecordDto);
    }

    @ApiOperation({
        summary: 'Obtiene todos los registros clinicos',
        description: 'Decuelve una lista de todos los registros clinicos, incluyendo los datos relacionados del paciente y tipo tumor'

    })

    @ApiResponse({
        status:200,
        description: 'Lista de registros clinicos devuelta exitosamente',
        type: [ClinicalRecord]
    })

    @Get()
    findAll(): Promise<ClinicalRecord[]>{
        return this.clinicalRecordService.findAll();
    }

    @Get(':id')
    @ApiOperation({
        summary: 'Obtiene un registro clinico por ID',
        description: 'Busca un registro clinico especifico de su ID numerico'
    })

    @ApiParam({
        name: 'id',
        description: ' ID numerico del registro clinico',
        type: String
    })

    @ApiResponse({
        status: 200,
        description: 'Registro clinico encontrado exitosamente',
        type: ClinicalRecord
    })
    @ApiResponse({
        status:404,
        description: 'Regiatro clinico no encontrado'
    })
    findById(@Param('id', ParseIntPipe) id:number): Promise<ClinicalRecord>{
        return this.clinicalRecordService.findById(id);
    }

    @Delete(':id')
    @HttpCode(HttpStatus.NO_CONTENT)

    @ApiOperation({
        summary:'Elimina un registro clinico',
        description: 'Elimina permanentemente un registro clinico por su ID'
    })

    @ApiParam({
        name:'id',
        description:'ID numerico del registro clinico a eliminar',
        type:String
    })

    @ApiResponse({
        status: 204,
        description:'Registro clinico eliminado exitosamente'
    })

    @ApiResponse({
        status: 404,
        description: 'Registro clinico no encontrado'
    })
    delete(@Param('id', ParseIntPipe) id: number): Promise<void>{
        return this.clinicalRecordService.delete(id);
    }
}

File: springMicro/src/main/java/com/geno/springGateway/clinicalRecord/infraestructure/controller/ClinicalRecordController.java
package com.geno.springGateway.clinicalRecord.infraestructure.controller;
import com.geno.springGateway.clinicalRecord.application.dto.UpdateClinicalRecordStageDto;
import com.geno.springGateway.clinicalRecord.application.dto.UpdateClinicalRecordTreatmentProtocolDto;
import com.geno.springGateway.common.ApiRestTemplateResponse;
import com.geno.springGateway.clinicalRecord.application.dto.ClinicalRecordInDto;
import com.geno.springGateway.clinicalRecord.application.dto.ClinicalRecordOutDto;
import com.geno.springGateway.clinicalRecord.application.service.ClinicalRecordService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

/**
 * @author mendez
 */
@RestController
@RequestMapping("/api/v1/clinical-records")//Modificarlo
@RequiredArgsConstructor
@Tag(name = "Historias Clínicas", description = "Operaciones CRUD sobre Historias Clínicas (Proxy a NestJS)")
public class ClinicalRecordController {

    private final ClinicalRecordService clinicalRecordService;

    @Operation(summary = "Obtener lista", description = "Recupera todas las Historias Clínicas.")
    @GetMapping
    public ResponseEntity<ApiRestTemplateResponse<List<ClinicalRecordOutDto>>> getAll() {
        List<ClinicalRecordOutDto> records = clinicalRecordService.findAll();
        return ResponseEntity.ok(new ApiRestTemplateResponse<>("Success", "Historias Clínicas encontradas", records));
    }

    @Operation(summary = "Obtener por ID", description = "Busca una Historia Clínica específica.")
    @ApiResponse(responseCode = "200", description = "Historia Clínica encontrada con éxito.")
    @ApiResponse(responseCode = "404", description = "Historia Clínica no encontrada (ID no existe).")
    @GetMapping("/{id}")
    public ResponseEntity<ApiRestTemplateResponse<ClinicalRecordOutDto>> getById(@PathVariable Long id) {
        ClinicalRecordOutDto dto = clinicalRecordService.findById(id);
        return ResponseEntity.ok(new ApiRestTemplateResponse<>("Success", "Historia Clínica encontrada", dto));
    }

    @Operation(summary = "Crear nuevo", description = "Registra una nueva Historia Clínica.")
    @ApiResponse(responseCode = "201", description = "Historia Clínica creada exitosamente.")
    @ApiResponse(responseCode = "400", description = "Solicitud inválida (error de validación en los datos de entrada, ej. ID de paciente o tumor vacío).")
    @PostMapping
    public ResponseEntity<ApiRestTemplateResponse<ClinicalRecordOutDto>> create(@RequestBody ClinicalRecordInDto dto) {
        ClinicalRecordOutDto created = clinicalRecordService.create(dto);
        return new ResponseEntity<>(new ApiRestTemplateResponse<>("Created", "Historia Clínica creada", created), HttpStatus.CREATED);
    }
    @Operation(summary = "Actualizar Etapa Clínica", description = "Actualiza solo el campo 'stage'.")
    @ApiResponse(responseCode = "200", description = "Etapa Clínica actualizada con éxito.")
    @ApiResponse(responseCode = "400", description = "Solicitud inválida (el campo está vacío o es nulo).")
    @ApiResponse(responseCode = "404", description = "Historia Clínica no encontrada (ID no existe).")
    @PatchMapping("/{id}/stage")
    public ResponseEntity<ApiRestTemplateResponse<ClinicalRecordOutDto>> patchStage(
            @PathVariable Long id,
            @RequestBody UpdateClinicalRecordStageDto dto)
    {
        ClinicalRecordOutDto patched = clinicalRecordService.patchStage(id, dto);
        return ResponseEntity.ok(new ApiRestTemplateResponse<>("Success", "Etapa clínica actualizada", patched));
    }

    @Operation(summary = "Actualizar Protocolo de Tratamiento", description = "Actualiza solo el campo 'treatmentProtocol'.")
    @ApiResponses({
            @ApiRestTemplateResponse(responseCode = "200", description = "Protocolo de tratamiento actualizado con éxito."),
            @ApiRestTemplateResponse(responseCode = "400", description = "Solicitud inválida (el campo está vacío o es nulo)."),
            @ApiRestTemplateResponse(responseCode = "404", description = "Historia Clínica no encontrada (ID no existe).")
    })
    @PatchMapping("/{id}/treatment-protocol")
    public ResponseEntity<ApiRestTemplateResponse<ClinicalRecordOutDto>> patchTreatment(
            @PathVariable Long id,
            @RequestBody UpdateClinicalRecordTreatmentProtocolDto dto)
    {
        ClinicalRecordOutDto patched = clinicalRecordService.patchTreatment(id, dto);
        return ResponseEntity.ok(new ApiRestTemplateResponse<>("Success", "Protocolo de tratamiento actualizado", patched));
    }

    @Operation(summary = "Eliminar", description = "Elimina una Historia Clínica por su ID.")
    @ApiResponses({
            @ApiRestTemplateResponse(responseCode = "200", description = "Historia Clínica eliminada con éxito."),
            @ApiRestTemplateResponse(responseCode = "404", description = "Historia Clínica no encontrada (ID no existe).")
    })
    @DeleteMapping("/{id}")
    public ResponseEntity<ApiRestTemplateResponse<Void>> delete(@PathVariable Long id) {
        clinicalRecordService.delete(id);
        return new ResponseEntity<>(new ApiRestTemplateResponse<>("Success", "Historia Clínica eliminada", null), HttpStatus.OK);
    }
}